File: tsd/tsd_cpp/core/types/trade.cpp
#include "trade.hpp"
#include <sstream>
#include <iomanip>
#include <random>
#include <chrono>

namespace tsd {
namespace core {

std::string to_string(TimeInForce tif) {
    switch (tif) {
        case TimeInForce::IOC: return "IOC";
        case TimeInForce::GTC: return "GTC";
        case TimeInForce::DAY: return "DAY";
        default: return "UNKNOWN";
    }
}

std::string to_string(OrderSide side) {
    switch (side) {
        case OrderSide::BUY: return "BUY";
        case OrderSide::SELL: return "SELL";
        default: return "UNKNOWN";
    }
}

std::string to_string(OrderType type) {
    switch (type) {
        case OrderType::LIMIT: return "LIMIT";
        case OrderType::MARKET: return "MARKET";
        default: return "UNKNOWN";
    }
}

TimeInForce time_in_force_from_string(const std::string& str) {
    if (str == "IOC") return TimeInForce::IOC;
    if (str == "GTC") return TimeInForce::GTC;
    if (str == "DAY") return TimeInForce::DAY;
    throw std::invalid_argument("Unknown TimeInForce: " + str);
}

OrderSide order_side_from_string(const std::string& str) {
    if (str == "BUY") return OrderSide::BUY;
    if (str == "SELL") return OrderSide::SELL;
    throw std::invalid_argument("Unknown OrderSide: " + str);
}

OrderType order_type_from_string(const std::string& str) {
    if (str == "LIMIT") return OrderType::LIMIT;
    if (str == "MARKET") return OrderType::MARKET;
    throw std::invalid_argument("Unknown OrderType: " + str);
}

ExternalOrder::ExternalOrder(Instrument instr, Exchange exchange, OrderSide side, OrderType type,
                           double price, std::optional<double> qty, std::optional<double> quote_qty,
                           TimeInForce tif, int64_t time_sent, const std::string& portfolio)
    : instr(instr), exchange(exchange), order_side(side), order_type(type),
      price(price), qty(qty), quote_order_qty(quote_qty),
      time_in_force(tif), time_sent(time_sent), portfolio(portfolio),
      client_order_id(generate_client_order_id(portfolio)) {
}

std::string ExternalOrder::generate_client_order_id(const std::string& portfolio) {
    // Generate UUID-like string
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> hex_dis(0, 15);

    std::stringstream ss;
    ss << portfolio << "_";

    // Generate 8 character hex string
    for (int i = 0; i < 8; ++i) {
        ss << std::hex << hex_dis(gen);
    }

    return ss.str();
}

} // namespace core
} // namespace tsd

File: tsd/core/types/trade.py
import uuid
from enum import StrEnum
from typing import Optional

import csp

from core.types.exchange import Exchange
from core.types.instrument import Instrument
from config.portfolio_config import Portfolio

MAX_CLIENT_ORDER_ID_LEN = 36
UUID_LEN = 8
DELIMITER_LEN = 1
MAX_PORTFOLIO_ID_LEN = MAX_CLIENT_ORDER_ID_LEN - UUID_LEN - DELIMITER_LEN


class TimeInForce(StrEnum):
    IOC = "IOC"
    GTC = "GTC"
    DAY = "DAY"


class OrderSide(StrEnum):
    BUY = "BUY"
    SELL = "SELL"


class OrderType(StrEnum):
    LIMIT = "LIMIT"
    MARKET = "MARKET"


class ExternalOrder(csp.Struct):
    instr: Instrument
    exchange: Exchange
    order_side: OrderSide
    order_type: OrderType

    price: float
    qty: Optional[float] = None  # optional: base 기반 주문용 BTCUSDT -> 0.1 BTC
    quote_order_qty: Optional[float] = None  # optional: quote 기반 주문용 BTCUSDT -> 1000 USDT

    time_in_force: TimeInForce
    time_sent: int
    portfolio: Portfolio

    # optional parameters
    # stop_price: float = 0.0                  # optional: stop order trigger price
    # post_only: bool = False                  # optional: maker only
    # reduce_only: bool = False                # optional: 포지션 감소 전용
    # margin_type: str = "cash"                # optional: spot은 'cash', margin은 cross/isolated
    # leverage: int = 1                        # optional: spot은 1배

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.client_order_id = generate_client_order_id(self.portfolio)


class Trade(csp.Struct):
    instr: Instrument
    price: float
    qty: float
    side: OrderSide


class ExecutionReport(csp.Struct):
    test: str


def generate_client_order_id(portfolio: Portfolio) -> str:
    """
    Generate a client_order_id with only portfolio metadata.
    Format: <portfolio>_<8char UUID>
    """
    validate_portfolio_id(portfolio)
    uid = uuid.uuid4().hex[:8]
    return f"{portfolio}_{uid}"


def validate_portfolio_id(portfolio: Portfolio):
    if len(portfolio) > MAX_PORTFOLIO_ID_LEN:
        raise ValueError(f"portfolio too long (max {MAX_PORTFOLIO_ID_LEN} characters): {portfolio}")
